File: form/validate.go
package form

import (
	"fmt"
	"log"

	"github.com/reggieanim/god-core/validator"
)

var kindMap = map[string]bool{
	"text":       true,
	"block":      true,
	"select":     true,
	"leftClick":  true,
	"rightClick": true,
	"saveState":  true,
	"loadState":  true,
	"nextPage":   true,
	"wait":       true,
	"condEval":   true,
	"eval":       true,
	"navigate":   true,
	"notify":     true,
}

func validate(ins map[string]interface{}) bool {
	_, desOk := ins["description"].(string)
	_, fildOk := ins["field"].(string)
	_, valOk := ins["value"].(string)
	kind, kindOk := ins["kind"].(string)
	_, doesNotMatchKind := kindMap[kind]
	v := validator.New()
	v.Check(desOk, "description", "Needs a 'description' property of string")
	v.Check(fildOk, "field", "Needs a 'field' property of 'text'")
	v.Check(valOk, "value", "Needs a 'value' property of string")
	v.Check(desOk, "description", "Needs a description property")
	v.Check(kindOk, "kind", "Needs a 'kind' property")
	v.Check(doesNotMatchKind, "noKind", "Needs a 'kind' property of 'text' || nextPage || select || leftClick || rightClick' || 'wait'")
	if !v.Valid() {
		for _, v := range v.Errors {
			fmt.Println(v)
		}
		log.Println("Could not validate")
	}
	return true
}

func validateEval(ins map[string]interface{}) bool {
	fmt.Println(ins)
	_, evalOk := ins["evalExpression"].(string)
	_, fieldOk := ins["field"].(string)
	v := validator.New()
	v.Check(evalOk, "evalOk", "Needs a 'evalExpression' property of string for extract")
	v.Check(fieldOk, "field", "Needs an 'field' property of string for extract")
	if !v.Valid() {
		for _, v := range v.Errors {
			fmt.Println(v)
		}
		log.Println("Could not validate")
	}
	return true
}

File: src/helpers/validation/validate.ts
import { FormInstructions } from "../../types/types";

const kindMap: { [key: string]: boolean } = {
  text: true,
  notify: true,
  select: true,
  leftClick: true,
  rightClick: true,
  wait: true,
  eval: true,
  condEval: true,
  block: true,
};

class Validator {
  errors: string[] = [];

  check(condition: boolean, field: string, message: string) {
    if (!condition) {
      this.errors.push(`${field}: ${message}`);
    }
  }

  valid(): boolean {
    return this.errors.length === 0;
  }
}

export const validate = (instructions: FormInstructions): boolean => {
  if (Object.keys(instructions).length === 0) return true;

  const isValidDescription = typeof instructions.description === "string";
  const isValidField = typeof instructions.field === "string";
  const isValidValue = typeof instructions.value === "string";
  const insKind = instructions.kind;
  const isValidKind = typeof insKind === "string";
  const doesNotMatchExpectedKind = insKind ? kindMap[insKind] : false;

  const v = new Validator();
  v.check(isValidDescription, "description", "Needs a 'description' property of string");
  v.check(isValidField, "field", "Needs a 'field' property of 'text'");
  v.check(isValidValue, "value", "Needs a 'value' property of string");
  v.check(isValidDescription, "description", "Needs a description property");
  v.check(isValidKind, "kind", "Needs a 'kind' property");
  v.check(
    doesNotMatchExpectedKind,
    "noKind",
    "Needs a 'kind' property of 'text' || condEval || select || leftClick || rightClick' || 'wait' || 'notify' || 'eval' || block"
  );

  if (!v.valid()) {
    for (const error of v.errors) {
      console.error(error);
    }
    console.log("Could not validate");
  }

  return v.valid();
};
